lib/puppet/pops/types/p_encrypted_type.rb

```diff
@@ -0,0 +1,297 @@
+# format is textual, and as binary content when a serialization format supports this.
+#
+# @api public
+class PEncryptedType < PAnyType
+
+  # Represents an Encrypted value
+  # @api public
+  #
+  class Encrypted
+    attr_reader :format
+    attr_reader :encrypted_key
+    attr_reader :crypt
+    attr_reader :encrypted_fingerprint
+
+    # Creates a new Encrypted
+    # This API is private - please use `Encrypted.encrypt` method to create an instance.
+    # This method is only accessible for serialization purposes.
+    #
+    # @api private
+    #
+    def initialize(format, encrypted_key, crypt, encrypted_fingerprint)
+      binary = Puppet::Pops::Types::PBinaryType::Binary
+      @format = format
+      @encrypted_key, @crypt, @encrypted_fingerprint = make_binary(encrypted_key, crypt, encrypted_fingerprint)
+    end
+
+    # @api private
+    def make_binary(*vars)
+      vars.map {|v| Puppet::Pops::Types::PBinaryType::Binary.from_binary(v) }
+    end
+    private :make_binary
+
+    # Presents the binary content as a base64 encoded string (without line breaks).
+    # While this method is public, the content of the String is not API and may change.
+    #
+    # @api public
+    #
+    def to_s
+      [ @format,
+        @encrypted_key,
+        @crypt,
+        @encrypted_fingerprint,
+        ].to_s
+    end
+
+    # @api public
+    #
+    def hash
+      # this is the encrypted random key which should give a reasonable spread
+      @encrypted_key.hash
+    end
+
+    # Returns identity equality since two Encrypted for the same value will always be different.
+    # @param o [Object] Any object to check for equality
+    # @return [Boolean]
+    # @api public
+    #
+    def eql?(o)
+      # Can only equal the same instance since each encryption is based on random values
+      # No need to check for anything besides object identity
+      __id__ == o.__id__
+    end
+
+    # Returns identity equality since two Encrypted for the same value will always be different.
+    # @param o [Object] Any object to check for equality
+    # @return [Boolean]
+    # @api public
+    #
+    def ==(o)
+      self.eql?(o)
+    end
+
+    # Encrypts data using a certificate, a cipher name and Any data - returns an Encrypted.
+    # This is the API for creating an `Encrypted` value.
+    #
+    # @param certificate [Puppet::SLL::Certificate] The certificate to use (i.e. for the receiver's public key)
+    # @param cipher [String] The name of the cipher to use - for example 'AES-256-CBC'
+    # @param data [Object] Any puppet rich data data type
+    # @return [Puppet::Pops::Types::PEncryptedType::Encrypted] An Encrypted puppet data type value
+    # @api public
+    #
+    def self.encrypt(certificate, cipher, data)
+      key = certificate.content.public_key
+      fingerprint = certificate.fingerprint
+      format = 'json,' + cipher
+      data = serialize(data)
+
+      aes_encrypt = OpenSSL::Cipher.new(cipher).encrypt
```

Comment:
This seems to indicate that AES is the only option. While this is a great default, users will expect to be able to pick anything from the underlying OpenSSL toolkit.

Also, an error message should be returned if an invalid cipher is chosen with the set of allowed ciphers.

Not sure how this is going to work if the server and client don't have compatible ciphers. Does a fact need to be delivered from the client and a set of ciphers be able to be passed here so that the usual TLS method of 'drop to strongest negotiated' will work?

Reply:
I can do the meaningful error bit right away. The finding of "strongest available" requires more - a fact plus perhaps giving an array of ciphers to the `Encrypted` - it picks the strongest. Any tips on how to find the strongest among a set of ciphers?

Reply:
@hlindberg Unfortunately, the 'correct' cipher needs to be up to the users. I would just mirror whatever `openssh` does as the default.